### PR TITLE
fix #8903 Encode workspacePath before being used as URL hash value. 

### LINKED ISSUE #8903: Open workspace doesn't work when folder name contain "%20"
### Bug Description:
When trying to open workspace of folder that its name contain "%20" no workspace will be opened.
I try to do the same scenario on vscode and the workspace was opened.

### Steps to Reproduce:

1. Create new folder "my%20test"
2. Select "File" ->"Open workspace..."  and choose folder "my%20test"
3. The workspaces doesn't open and we get the following screen: 
image

### Additional Information

- Operating System: gitpot
- Theia Version: latest

### PR DESCRIPTION
**Pull request by shuyaqian**
#### What it does

fixes [#8903](https://github.com/eclipse-theia/theia/issues/8903)
Encode workspacePath before being used as URL hash value when openNewWindow.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Review by vince-fugnitto — COMMENTED** [review]
Thank you for your contribution @shuyaqian :+1: 
Before it can be accepted, please be sure to:
- include a proper [sign-off](https://github.com/eclipse-theia/theia/blob/master/CONTRIBUTING.md#sign-your-work) for your commit.
- be sure to sing the [Eclipse Contributor Agreement (ECA)](https://github.com/eclipse-theia/theia/blob/master/CONTRIBUTING.md#sign-your-work) with the same email as your `sign-off`

This is the reason for the failing ECA check.

**Comment by shuyaqian** [comment]
> Thank you for your contribution @shuyaqian 👍
> Before it can be accepted, please be sure to:
> 
> * include a proper [sign-off](https://github.com/eclipse-theia/theia/blob/master/CONTRIBUTING.md#sign-your-work) for your commit.
> * be sure to sing the [Eclipse Contributor Agreement (ECA)](https://github.com/eclipse-theia/theia/blob/master/CONTRIBUTING.md#sign-your-work) with the same email as your `sign-off`
> 
> This is the reason for the failing ECA check.



> Thank you for your contribution @shuyaqian 👍
> Before it can be accepted, please be sure to:
> 
> * include a proper [sign-off](https://github.com/eclipse-theia/theia/blob/master/CONTRIBUTING.md#sign-your-work) for your commit.
> * be sure to sing the [Eclipse Contributor Agreement (ECA)](https://github.com/eclipse-theia/theia/blob/master/CONTRIBUTING.md#sign-your-work) with the same email as your `sign-off`
> 
> This is the reason for the failing ECA check.

I've added it. Thank you.

**Review by vince-fugnitto — COMMENTED** [review]
@shuyaqian the changes work for me when starting the example `browser` application, but not the `electron` (desktop) application.

 To test you can do the following:
- start the example-electron app: `yarn start:electron`
- attempt to open a folder with `%2` in the name

<div align='center'>

image

</div>

**Comment by kittaakos** [comment]
What about this location?

https://github.com/eclipse-theia/theia/blob/f1d469033cfebb19d9e54215c9b746d133bd0206/packages/workspace/src/browser/workspace-service.ts#L153

**Comment by kittaakos** [comment]
> but not the `electron` (desktop) application.

I guess here is the missing part:

https://github.com/eclipse-theia/theia/blob/f1d469033cfebb19d9e54215c9b746d133bd0206/packages/core/src/electron-main/electron-main-application.ts#L275

This is quite a breaking, isn't it? I might be overlooking something, but if this is a breaking change, please include it in the changelog. Thanks!

**Comment by shuyaqian** [comment]
> > but not the `electron` (desktop) application.
> 
> I guess here is the missing part:
> 
> https://github.com/eclipse-theia/theia/blob/f1d469033cfebb19d9e54215c9b746d133bd0206/packages/core/src/electron-main/electron-main-application.ts#L275
> 
> This is quite a breaking, isn't it? I might be overlooking something, but if this is a breaking change, please include it in the changelog. Thanks!

Thank you for your suggestion. I've revised it.

**Review by vince-fugnitto — COMMENTED** [review]
@shuyaqian sorry for taking a while to get back to the pull-request, would you mind rebasing the pull-request to the latest master, and also including an breaking change note in the `changelog` regarding the changes (it might be better to be transparent that urls are now encoded when opening the application from now on).

**Comment by kittaakos** [comment]
> What about this location?
> 
> https://github.com/eclipse-theia/theia/blob/f1d469033cfebb19d9e54215c9b746d133bd0206/packages/workspace/src/browser/workspace-service.ts#L153

Please also check if we need to change it here. Thanks!

**Comment by shuyaqian** [comment]
> > What about this location?
> > https://github.com/eclipse-theia/theia/blob/f1d469033cfebb19d9e54215c9b746d133bd0206/packages/workspace/src/browser/workspace-service.ts#L153
> 
> Please also check if we need to change it here. Thanks!

I've revised it, thank you very much!

**Review by vince-fugnitto — APPROVED** [review]
@shuyaqian thank you for the changes! :+1: 
I confirmed that opening a workspace named `a%20b` can now be successfully opened in both the _browser_ and _electron_ targets.

File: packages/core/src/electron-main/electron-main-application.ts [hunk 1]
```diff
@@ -279,7 +279,7 @@ export class ElectronMainApplication {
             if (workspacePath === undefined) {
                 await this.openDefaultWindow();
             } else {
-                await this.openWindowWithWorkspace(workspacePath);
+                await this.openWindowWithWorkspace(encodeURI(workspacePath));
```
Comment (marechal-p) [inline packages/core/src/electron-main/electron-main-application.ts:282]: Encoding should be done when setting `withFragment`, not here.

**Comment by marechal-p** [comment]
@shuyaqian are this changes not useful anymore? Note that I had an unresolved comment.

**Pull request closed without merging**